controllers/progressiverollout_controller.go

```diff
@@ -60,24 +62,63 @@ func (r *ProgressiveRolloutReconciler) Reconcile(ctx context.Context, req ctrl.R
 		log.Error(err, "unable to fetch ProgressiveRollout")
 		return ctrl.Result{}, client.IgnoreNotFound(err)
 	}
-	// Always log the ApplicationSet owner
+
 	log = r.Log.WithValues("applicationset", pr.Spec.SourceRef.Name)
 
 	for _, stage := range pr.Spec.Stages {
 		log = r.Log.WithValues("stage", stage.Name)
 
-		targets, err := r.GetTargetClusters(stage.Targets.Clusters.Selector)
+		// Get the clusters to update
+		clusters, err := r.getClustersFromSelector(stage.Targets.Clusters.Selector)
 		if err != nil {
-			log.Error(err, "unable to fetch targets")
+			log.Error(err, "unable to fetch clusters")
+			return ctrl.Result{}, err
+		}
+		r.Log.V(1).Info("clusters selected", "clusters", fmt.Sprintf("%v", clusters.Items))
+
+		// Get the Applications owned by the ProgressiveRollout targeting the clusters
+		apps, err := r.getOwnedAppsFromClusters(clusters, pr)
+		if err != nil {
+			log.Error(err, "unable to fetch apps")
+			return ctrl.Result{}, err
+		}
+		r.Log.V(1).Info("apps selected", "apps", fmt.Sprintf("%v", apps))
+
+		// Remove the annotation from the OutOfSync Applications before passing them to the Scheduler
+		// This action allows the Scheduler to keep track at which stage an Application has been synced.
+		outOfSyncApps := utils.FilterAppsBySyncStatusCode(apps, argov1alpha1.SyncStatusCodeOutOfSync)
+		if err = r.removeAnnotationFromApps(&outOfSyncApps, utils.ProgressiveRolloutSyncedAtStageKey); err != nil {
 			return ctrl.Result{}, err
 		}
-		r.Log.V(1).Info("targets selected", "targets", targets.Items)
-		r.Log.Info("stage completed")
+
+		// Get the Applications to update
+		scheduledApps := scheduler.Scheduler(apps, stage)
+
+		for _, s := range scheduledApps {
+			// TODO: add sync method here
+			r.Log.Info("syncing app", "app", s)
+		}
+
+		if scheduler.IsStageFailed(apps, stage) {
+			// TODO: updated status
+			r.Log.Info("stage failed")
+			return ctrl.Result{}, nil
+		}
+
+		if scheduler.IsStageComplete(apps, stage) {
+			// TODO: update status
+			r.Log.Info("stage completed")
```

Comment:
it's a small detail, but for all those logs is there a way to log the name of the stage? I realise it's at the top of the function, but when there are several goroutines looking after individual progressive rollouts, it'll become quite difficult to read?

Reply:
My understanding is that every log line after https://github.com/Skyscanner/argocd-progressive-rollout/pull/25/files#diff-b1a5cf22f2be07477047c7c268c9a2077f75c3fdadc285eb862dab8bc3aaad31R69 will have the stage name. Am I missing something?

Reply:
nope, it was me, i didn't realise the `r.Log` would already include the stage name with `WithValues`